Metrics documentation (DB-185) (#3884)
EventStore/EventStore: docs/metrics.md

# Metrics

> Since EventStoreDB version `v23.6.0`, some metrics in [prometheus format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format) are available on the `/metrics` endpoint. Promethus can be configured to scrape this endpoint directly. The metrics are configured in `telemetryconfig.json`. In the near future we hope to publish a sample grafana dashboard.

Reviewer: ```suggestion
EventStoreDB collects metrics in [prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/#exposition-formats) , available on the `/metrics` endpoint. Promethus can be configured to scrape this endpoint directly. The metrics are configured in `telemetryconfig.json`. 
```
Resoning for the change: 
- the doc are versionend, so this will appear in 23.6 version and the following one. no need to mention "since"
- prefer to link to the online doc of prometheus ( so that _they_ can track that we use their format ) 
- remove "in the near future" : because it's a roadmap promise in the documentation .

Reviewer 2: thanks Yves, most of those are my doing :)
- I agree about the link, thanks!
- I see what you mean about the promise, do you think it would be ok to say 'in the future'? im happy if you'd prefer to leave it out entirely
- I mentioned 23.6 because there is sort of some overlap with the older prometheus exporter and I wanted to make it clear somehow that this one is the new one (the old one is still documented).

Reviewer: >  see what you mean about the promise, do you think it would be ok to say 'in the future'? im happy if you'd prefer to leave it out entirely

I'd leave it out .

> 23.6 : 
Let's amend the prometheus doc then ?
add a link to this , and make it clear it's about the Stats/gossip endopint

::: note
`/metrics` does not yet contain metrics for Projections and Persistent Subscriptions. To view these in Prometheus it is still recommended to use the [Prometheus exporter](./diagnostics.md#prometheus)
:::

## Metrics Reference

### Caches

#### Cache hits and misses
EventStoreDB tracks cache hits/misses metrics for `stream-info` and `chunk` caches.

| Time Series | Type       | Description |
| :---------- | :--------- | :---------- |

> | `eventstore_cache_hits_misses{cache=<CACHE_NAME>,kind=<"hits"\|"misses">}` | [Counter](#common-types) | Total hits/misses on _CACHE_NAME_ cache   |

Reviewer: ```suggestion
| `eventstore_cache_hits_misses{cache=<CACHE_NAME>,kind=<"hits"\|"misses">}` | [Counter](https://prometheus.io/docs/tutorials/understanding_metric_types/) | Total hits/misses on _CACHE_NAME_ cache   |
```
Reasoning: either link directly to the source doc in prometheus, or copy some information in that section .
the text when following that link is " go to prometheus doc" without a link to it .

| :---------- | :--------- | :---------- |
| `eventstore_checkpoints{name=<CHECKPOINT_NAME>,read="non-flushed"}` | [Gauge](#common-types) | Value for _CHECKPOINT_NAME_ checkpoint|

Example Configuration:
```
"Checkpoints": {
	"Replication": true,
	"Chaser": false,
	"Epoch": false,
	"Index": false,
	"Proposal": false,
	"Truncate": false,
	"Writer": false,
	"StreamExistenceFilter": false
}
```

Example Output:
```
# TYPE eventstore_checkpoints gauge
eventstore_checkpoints{name="replication",read="non-flushed"} 613363 1688054162478
```

### Events

These metrics track events written to and read from the server, including reads from caches.

| Time Series | Type       | Description |
| :---------- | :--------- | :---------- |
| `eventstore_io_bytes{activity="read"}`               | [Counter](#common-types) | Event bytes read |
| `eventstore_io_events{activity=<"read"\|"written">}` | [Counter](#common-types) | Events read/written |

Example Configuration:
```
"Events": {
	"Read": false,
	"Written": true
}
```

Example Output:
```
# TYPE eventstore_io_events counter
# UNIT eventstore_io_events events
eventstore_io_events{activity="written"} 320 1687963622074
```

### Gossip

> Usually a node pushes new gossip to other nodes periodically or when its view of the cluster changes. Sometimes nodes pull gossip from each other if there is a suspected network problem. We measure round trip latency and processing time.

Reviewer: ```suggestion
Measures the round trip latency and processing time of gossip.
Usually a node pushes new gossip to other nodes periodically or when its view of the cluster changes.  Sometimes nodes pull gossip from each other if there is a suspected network problem. 
```
reasoning: tell fisrt _what_ is measured, then why .

| `eventstore_current_incoming_grpc_calls`                   | [Gauge](#common-types) | Inflight gRPC calls i.e. gRPC requests that have started on the server but not yet stopped |
| `eventstore_incoming_grpc_calls{kind="total"}`             | [Counter](#common-types) | Total gRPC requests served                           |
| `eventstore_incoming_grpc_calls{kind="failed"}`            | [Counter](#common-types) | Total gRPC requests failed                           |
| `eventstore_incoming_grpc_calls{kind="unimplemented"}`     | [Counter](#common-types) | Total gRPC requests made to unimplemented methods    |
| `eventstore_incoming_grpc_calls{kind="deadline-exceeded"}` | [Counter](#common-types) | Total gRPC requests for which deadline have exceeded |

Example Configuration:
```
"IncomingGrpcCalls": {
	"Current": true,
	"Total": false,
	"Failed": true,
	"Unimplemented": false,
	"DeadlineExceeded": false
}
```

Example Output:
```
# TYPE eventstore_current_incoming_grpc_calls gauge
eventstore_current_incoming_grpc_calls 1 1687963622074

# TYPE eventstore_incoming_grpc_calls counter
eventstore_incoming_grpc_calls{kind="failed"} 1 1687962877623
```

> #### gRPC methods

Reviewer: ```suggestion
#### Client protocol gRPC methods
```

> In addition, EventStoreDB also records metrics for each of the gRPC methods : `StreamRead`, `StreamAppend`, `StreamBatchAppend`, `StreamDelete` and `StreamTombstone`. They are grouped together according to the mapping defined in the configuration.

Reviewer: ```suggestion
In addition, EventStoreDB also records metrics for each of client protocol gRPC methods : `StreamRead`, `StreamAppend`, `StreamBatchAppend`, `StreamDelete` and `StreamTombstone`. They are grouped together according to the mapping defined in the configuration.
```

	"StreamTombstone": ""
}
```

Example Output:
```
# TYPE eventstore_grpc_method_duration_seconds histogram
# UNIT eventstore_grpc_method_duration_seconds seconds
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="1E-06"} 0 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="1E-05"} 0 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="0.0001"} 129 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="0.001"} 143 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="0.01"} 168 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="0.1"} 169 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="1"} 169 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="10"} 169 1688157491029
eventstore_grpc_method_duration_seconds_bucket{activity="append",status="successful",le="+Inf"} 169 1688157491029
```

> ### Kestrel

Reviewer: maybe a link to what Kesterl is migh be good ?

Author: Yup. Added it!

| Time Series | Type       | Description |
| :---------- | :--------- | :---------- |
| `eventstore_kestrel_connections`  | [Gauge](#common-types) | Number of open kestrel connections |

Example Configuration:
```
"Kestrel": {
	"ConnectionCount": true
}
```

Example Output:
```
# TYPE eventstore_kestrel_connections gauge
eventstore_kestrel_connections 1 1688070655500
```

### Process

> EventStoreDB collects large range of metrics for its own process.

Reviewer: ```suggestion
EventStoreDB collects large range of metrics about the running process.
```

::: note
Enabling `Queues.Processing` can cause a lot more time series to be generated, according to the `QueueLabels` and `MessageTypes` configuration.
:::

Example Configuration:
```
Queues": {
	"Busy": true,
	"Length": true,
	"Processing": false
}

"QueueLabels": [
	{
		"Regex": "StorageReaderQueue #.*",
		"Label": "Readers"
	},
	{
		"Regex": ".*",
		"Label": "Others"
	}
]
```

Example Output:
```
# TYPE eventstore_queue_busy_seconds counter
# UNIT eventstore_queue_busy_seconds seconds
eventstore_queue_busy_seconds{queue="Readers"} 1.04568158125 1688157491029
eventstore_queue_busy_seconds{queue="Others"} 0 1688157491029

# TYPE eventstore_queue_queueing_duration_max_seconds gauge
# UNIT eventstore_queue_queueing_duration_max_seconds seconds
eventstore_queue_queueing_duration_max_seconds{name="Readers",range="16-20 seconds"} 0.06434454 1688157489545
eventstore_queue_queueing_duration_max_seconds{name="Others",range="16-20 seconds"} 0 1688157489545
```

### Status

> EventStoreDB can track the current status of the `Node`, the `Index`, and the `Scavenge`

Reviewer: ```suggestion
EventStoreDB tracks the current status of the `Node` role  as well as the `Index`, and the `Scavenge` processes.
```
I'd put that one on top , this is probably yhe first one we want people to actually track

Author: Currently, metrics sections are arranged alphabetically. Please let me know if you think some other order is better. (I personally prefer alphabetical order.)

Author: Other then that, I have rephrased the statement according to your suggestion.

| Time Series | Type       | Description |
| :---------- | :--------- | :---------- |
| `eventstore_statuses{name=<NAME>,status=<STATUS>}` | [Gauge](#common-types) | Number of seconds since the 1970 epoch when _NAME_ most recently had the status _STATUS_ |

For a given _NAME_, the current status can be determined by taking the max of all the time series with that name.

`Index` can have one of the following statuses:
- Opening (loading/verifying the PTables)
- Rebuilding (indexing previously written records on start up)
- Initializing (initializing any other parts of the index e.g. StreamExistenceFilter on start up)
- Merging
- Scavenging
- Idle

`Scavenge` can have one of the following statuses:
- Accumulation
- Calculation
- Chunk Execution
- Chunk Merging
- Index Execution
- Cleaning
- Idle

> `Node` can have one of the node statuses: `Leader`, `Follower`, etc.

Reviewer: ```suggestion
Read more about the node status [here](cluster.html#node-roles)
```

Example Configuration:
```
"Statuses": {
	"Index": true,
	"Node": false,
	"Scavenge": false
}
```

Example Output:
```
# TYPE eventstore_checkpoints gauge
eventstore_statuses{name="Index",status="Idle"} 1688054162 1688054162477
```

### Storage Writer

| Time Series | Type       | Description |
| :---------- | :--------- | :---------- |
| `eventstore_writer_flush_size_max{range=<RANGE>}`             | [RecentMax](#recentmax) | Recent maximum flush size in bytes       |
| `eventstore_writer_flush_duration_max_seconds{range=<RANGE>}` | [RecentMax](#recentmax) | Recent maximum flush duration in seconds |

Example Configuration:
```
"Writer": {
	"FlushSize": true,
	"FlushDuration": false
}
```

Example Output:
```
# TYPE eventstore_writer_flush_size_max gauge
eventstore_writer_flush_size_max{range="16-20 seconds"} 410 1688056823193
```

> ### System

Reviewer: put this one near the `Process` as they are closely related

Reviewer 2: currently its alphabetical, i'd probably keep it that way but open to persuasion

Reviewer: is ok alphabetical
